docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,436 @@
+    uint32_t setting_count,
+    oe_enclave_t** enclave)
+  {
+      return oe_create_enclave(
+               path,
+               type,
+               flags,
+               settings,
+               setting_count,
+               __foo_ocall_function_table,
+               3,
+               __foo_ecall_info_table,
+                3,
+               enclave);
+  }
+  ```
+  - During the initialization of the local table, the host uses the following logic
+    to get the global id by name.
+  ```
+  uint32_t i, index;
+  for (i = 0; i < ecall_info_table_size; i++)
+  {
+    if found name in the global_table
+      index = the index of the name
+    else /* not found */
+      add the name to the table
+      index = the index of the name
+
+    /* Assign the ECALL id to the local table based on the global id. */
+    enclave->ecall_id_table[index] = i;
+  }
+  ```
+  - Modify the internal `oe_call_enclave_function` function to pass a global id
+    and the name string of the ECALL instead of hard-coded ECALL id.
+    The `global_id` is defined as a static variable. Given the same `global_id`
+    is used to retrieve the same ECALL across enclaves, the host initializes the
+    `global_id` (based on the name) in the first invocation and uses the cached value
+    for the subsequent ones.
+  ```
+  static uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL;
+  ...
+  /* Call enclave function. */
+  if ((_result = oe_call_enclave_function(
+           enclave,
+           &global_id,
+           __foo_ecall_info_table[foo_fcn_id_ecall_common_1_ecall].name,
+           _input_buffer,
+           _input_buffer_size,
+           _output_buffer,
+           _output_buffer_size,
+           &_output_bytes_written)) != OE_OK)
+      goto done;
+  ```
+
+In our previous example, assuming that the host instantiates the `foo` before `bar`.
+By applying the above scheme, the host will maintain the following data structures.
+- `_ecall_table`
+  ```
+  _ecall_table[0]: { "ecall_common_1_ecall" }
+  _ecall_table[1]: { "ecall_common_2_ecall_1" }
+  _ecall_table[2]: { "ecall_common_2_ecall_2" }
+  _ecall_table[3]: { "ecall_foo_ecall" }
+  _ecall_table[4]: { "ecall_bar_ecall" }
+  ```
+- `foo` (at the time `bar` is not initialized yet)
+  ```
+  enclave->ecall_id_table[0]: { 0 /* ecall_common_1_ecall */ }
+  enclave->ecall_id_table[1]: { 1 /* ecall_common_2_ecall_1 */ }
+  enclave->ecall_id_table[2]: { 2 /* ecall_common_2_ecall_2 */ }
+  enclave->ecall_id_table[3]: { 3 /* ecall_foo_ecall */ }
+  ```
+- `bar`
+  ```
+  enclave->ecall_id_table[0]: { 2 /* ecall_common_1_ecall */ }
+  enclave->ecall_id_table[1]: { 0 /* ecall_common_2_ecall_1 */ }
+  enclave->ecall_id_table[2]: { 1 /* ecall_common_2_ecall_2 */ }
+  enclave->ecall_id_table[3]: { OE_ECALL_ID_NULL /* ecall_foo_ecall */ }
+  enclave->ecall_id_table[4]: { 3 /* ecall_bar_ecall */ }
+  ```
+As we can see, the scheme allows the host to use the same global id to look up different ECALL id
```

Comment:
I didn't see why we need a global id at all.  Above said "both `foo_u.c` and `bar_u.c` will implement same wrapper functions of the three imported functions" which seems to be the issue, but you can solve that by edger8r prefixing the wrappers with foo and bar respectively and then I think you don't need a global id and the solution is simpler.  Or am I missing something?

Reply:
The downside of prefixing is that it would apply to every ecalls. This complicates the user experience; i.e., users need to add the prefix for both system ecalls (e.g., `foo_oe_log_init_ecall`, `bar_oe_log_init_ecall`) and their own ecalls (e.g., `foo_ecall_foo`, `bar_ecall_bar`).

This proposed design is to avoid making the prefix as a hard requirement when importing system edls.

@anakrish can you also comment on this?

Reply:
Yes, the user calls the enclave functions via the host-side wrapper functions. 
Given
```c
foo.edl

trusted {
   void ecall1();
};
```

Currently, the user invokes the ecall via:
```c
ecall1(enclave_handle);
```
Adding a prefix to the untrusted wrapper would change the user experience, and the user would have to make the ecall via:
```c
foo_ecall1(enclave_handle);
```

Intel SDK makes prefixing a user-choice via the `--use-prefix` edger8r option. Therefore the user can decide whether to use prefix or not (to avoid conflict), by passing --use-prefix to conflicting edls.

However the prefixing strategy does not work for system edls or user-written libraries.
Consider a reusable library (user_lib.a), that internally makes some ecalls. The library has an associated edl user_lib.edl:

```c
user_lib.edl:
trusted {
   void lib_ecall1();    // Function index is 0
};

user_lib.c :
void lib_function(enclave_handle h)
{
   ... do some stuff
   // call enclave function necessary for this library
   lib_ecall1(h);   <----- what prefix should this call use?
   // As written, the above wrapper will use index 0.
}
```
The redistributable library has to make the ecall at a specific location in code, and therefore it has to lockdown the prefix (which may be empty).

Now consider two enclaves that reuse this library:
```c

enclave1.edl
{
    import "user_lib.edl"; // lib_ecall1 has index zero
};

enclave2.edl
{
   import "something_else.edl";
   import "user_lib.edl";   <--- the function index for lib_ecall1 is different due to different imports.
}
```
Even if you make different prefixes for enclave1 and enclave2, the user_lib.a library has already chosen a specific prefix (in this case empty).
Therefore when the library is invoked via lib_function, passing in either enclave1 or enclave2 handles, the call to the wrapper `lib_ecall1` may fail since it may use a wrong index.

This is the scenario where prefixes wont work. If the ecall happens in a reusable library or the SDK itself, the call-site is bound to the prefix/index.

Reply:
@mingweishih @dthaler
I have created reproduction steps for the Intel SDK here:
https://github.com/anakrish/imported-ecall-issue

Reply:
@dthaler Does the explanation above make sense to you?

Reply:
@dthaler [logging.edl](https://github.com/openenclave/openenclave/blob/master/include/openenclave/edl/logging.edl) is another example that the prefix won't work.

 This system edl defines an ecall `oe_log_init_ecall`, which is used by the oehost library as follows.
https://github.com/openenclave/openenclave/blob/2a3da923e893e4263637f27983deb1016c7f8710/host/sgx/create.c#L1011

If user applies the prefix, say `foo`, then the oehost cannot correctly invoke the oeedger8r-generated implementation (in this case `foo_oe_log_init_ecall`). Instead, the default implementation will be used, which returns `OE_UNSUPPORTED`.

Reply:
Yes that makes sense to me, thanks!  See updated comment.